### PR TITLE
Allow 1->N broadcasts at the beginning and end to be fused

### PR DESCRIPTION
**Pull request by jamesr66a**

### CODE REVIEW

File: torch/csrc/jit/passes/onnx/peephole.cpp [hunk 1]
```diff
@@ -19,20 +19,34 @@ bool isBroadcasting(Node *node) {
...
+
+  ssize_t f = from_dim_end;
+  ssize_t t = to.size() - 1;
+  for (; f >= from_dim_start && t >= 0; --f, --t) {
+    if (from[f] != to[t]) return false;
+  }
+
+  // In the case that the 'to' tensor has leading ones in the same place that
+  // the 'from' tensor does, f will be less than from_dim_start rather than
+  // strictly equal. E.x.: to := [5, 1, 768] and from := [1, 1, 768]
```
Comment (jamesr66a) [inline torch/csrc/jit/passes/onnx/peephole.cpp]: @bddppq double check me here. I don't know if this affects your logic in C2 as well

File: torch/csrc/jit/passes/onnx/peephole.cpp [hunk 2]
```diff
@@ -19,20 +19,34 @@ bool isBroadcasting(Node *node) {
...
-// the dimension sizes must either be equal, or one of them does not exist.
+// First iterate over the 'from' tensor sizes. Ignore all leading and trailing
+// dimensions that are simply one, since they can be trivially broadcasted.
+// When iterating over the dimension sizes (with reduced 'from' tensor),
+// starting at the trailing dimension, the dimension sizes must either be equal,
+// or one of them does not exist.
 //
 //  equivalently:
 //
 // Test that 'from' is a suffix of 'to'.
```
Comment (ezyang) [inline torch/csrc/jit/passes/onnx/peephole.cpp]: This part of the comment is out of date now.

File: torch/csrc/jit/passes/onnx/peephole.cpp [hunk 3]
```diff
@@ -19,20 +19,34 @@ bool isBroadcasting(Node *node) {
   return broadcasting.count(node->kind());
 }
 
-// When iterating over the dimension sizes, starting at the trailing dimension,
-// the dimension sizes must either be equal, or one of them does not exist.
+// First iterate over the 'from' tensor sizes. Ignore all leading and trailing
+// dimensions that are simply one, since they can be trivially broadcasted.
+// When iterating over the dimension sizes (with reduced 'from' tensor),
+// starting at the trailing dimension, the dimension sizes must either be equal,
+// or one of them does not exist.
```
Comment (ezyang) [inline torch/csrc/jit/passes/onnx/peephole.cpp]: It's worth calling out that these semantics are NOT numpy style broadcasting, and they are not the generalized broadcasting that PyTorch implements in general.

**Comment by ezyang** [comment]
Is this going to be a Caffe2-only-ism (it's OK for this to be the case, but I doubt ONNX is going to adopt this logic for its broadcasting... so if it is caffe2 only logic, we should be clear about it.)

File: torch/csrc/jit/passes/onnx/peephole.cpp (hunk 1 above)
Comment (bddppq) [reply torch/csrc/jit/passes/onnx/peephole.cpp]: The logic looks good to me. But I think you need to check `ndim(from) <= ndim(to)`, e.g. for `to :=  [5, 768], from := [1, 1, 768]` it should return `false`.

**Comment by dzhulgakov** [comment]
Yes, it's Caffe2-only-ism, but it's a subset of numpy broadcasting and thus is reasonable. I think we want to enable full numpy broadcasting in ONNX, so this is a sub-step in that direction

**Pull request merged**